### PR TITLE
refactor(core): replace StorageProvider with ReportGenerator protocol

### PR DESCRIPTION
**Pull request by quanru**
## Summary

- Replace `StorageProvider` architecture with a protocol-based `ReportGenerator` that decouples report generation from `Agent` and `GroupedActionDump`
- Simplify `ScreenshotItem`: remove provider dependency, sync `create()`, replace `getData()` with `.base64` getter, use compact `{ $screenshot: id }` serialization format
- Implement two report modes with O(1) peak memory and incremental disk I/O:
  - **Inline mode**: truncate+append strategy — image tags written once, dump JSON truncated and rewritten
  - **Directory mode**: screenshots as PNG files, HTML overwritten per update
- Eliminate redundant `dump.serialize()` call in `onTaskUpdate` by caching and passing the serialized string through `writeOutActionDumps`
- Delete `storage/` directory (`provider.ts`, `memory.ts`, `file.ts`, `index.ts`) and unused `report-writer.ts`

## Performance impact

With 94 test cases sharing one Agent (dump accumulates):

| Metric | Before | After |
|--------|--------|-------|
| `ScreenshotItem.toSerializable()` | `{ base64: "..." }` (~200KB) | `{ $screenshot: "uuid" }` (~50B) |
| Dump size at test #94 | ~18MB | ~188KB |
| Serializations per `emitOnTaskUpdate` | 2× | 1× |
| Per-aiAssert serialization volume | ~180MB | ~940KB |

## Test plan

- [x] `npx nx build @midscene/core` passes
- [x] `npx nx test @midscene/core` — 498 passed (1 pre-existing failure in `utils.test.ts`)
- [x] `npx nx test @midscene/web` — 238 passed (5 pre-existing failures in `playwright-reporter.test.ts`)
- [x] New `report-generator.test.ts` — 15 tests covering inline/directory modes, deduplication, truncate+append, linear growth, and memory efficiency guarantees

### CODE REVIEW

File: packages/core/src/storage/file.ts [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    const filePath = this.registry.get(id);
+    if (filePath && fs.existsSync(filePath)) {
+      try {
+        fs.unlinkSync(filePath);
+      } catch {
+        // Ignore deletion errors
+      }
+    }
+    this.registry.delete(id);
+  }
```
[the remark this answers is not in the document]
Comment (quanru) [reply packages/core/src/storage/file.ts]: This is an intentional design decision. The async signatures are required to implement the `StorageProvider` interface, but we use synchronous operations internally for atomicity and simplicity.

**Rationale:**
1. **Interface compliance**: The `StorageProvider` interface defines async methods to support both sync and async implementations (e.g., MemoryStorage vs FileStorage).
2. **Atomicity**: Synchronous file operations ensure atomic writes, which is critical for screenshot data integrity.
3. **Performance**: For sequential screenshot storage operations, the overhead of async I/O scheduling outweighs the benefits. The comment on line 55-56 explicitly documents this decision.

The hybrid approach is intentional and provides a good balance between interface consistency and implementation efficiency.

File: packages/core/src/storage/file.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+  getDirectory(): string {
+    return this.directory;
+  }
+
+  get size(): number {
+    return this.registry.size;
+  }
+}
```
[the remark this answers is not in the document]
Comment (quanru) [reply packages/core/src/storage/file.ts]: Agreed, this is a valid improvement suggestion. However, I'd like to note:

**Current state:**
- The StorageProvider implementations (MemoryStorage and FileStorage) are relatively simple and have been tested through integration tests in the agent and reporter workflows.
- The core functionality (store/retrieve/delete) has been exercised in real-world scenarios during development.

**Future work:**
- Adding comprehensive unit tests for StorageProvider is definitely on the roadmap and would improve confidence in edge cases.
- This is a good candidate for a follow-up PR focused on testing infrastructure.

For this PR, the primary goal was to establish the storage abstraction and integrate it with the existing codebase. I'll create a separate issue to track adding comprehensive unit tests for the storage providers.

File: packages/web-integration/src/playwright/ai-fixture.ts [hunk 3]
```diff
@@ -109,16 +109,34 @@ export const PlaywrightAiFixture = (options?: {
...
+          } finally {
+            // Clean up
+            pageTempFiles.delete(idForPage);
+            pageAgentMap[idForPage]?.destroy();
+            delete pageAgentMap[idForPage];
+          }
+        })().catch((error) => {
+          console.error('Error in page close handler:', error);
+        });
       });
```
[the remark this answers is not in the document]
Comment (quanru) [reply packages/web-integration/src/playwright/ai-fixture.ts:118]: This is an intentional fire-and-forget pattern, and the race condition concern is already addressed by the reporter's synchronization mechanism.

**Why fire-and-forget:**
- The `page.on('close')` event handler cannot be async - it must be synchronous to avoid blocking the page close event.
- We wrap the async logic in an IIFE with `.catch()` for error handling (lines 137-139).

**Race condition mitigation:**
The reporter handles this gracefully:
1. The async operation writes to a temp file (line 123)
2. The reporter's `onEnd()` method waits for all pending report updates using `Promise.all()` (reporter/index.ts:193-200)
3. Even if the temp file write completes after `onTestEnd()`, the reporter has fallback cleanup logic (lines 202-217)

**Error handling:**
- Any errors in serialization are caught and logged (line 130)
- The cleanup in the `finally` block always executes (lines 131-136)

The current implementation balances event handler constraints with proper synchronization at the reporter level.

File: packages/visualizer/src/component/blackboard/index.tsx [hunk 4]
```diff
@@ -102,7 +102,16 @@ export const Blackboard = (props: {
...
+  const [screenshotBase64, setScreenshotBase64] = useState<string>('');
+
+  useEffect(() => {
+    const screenshotData = screenshot as unknown as
+      | { base64: string }
+      | undefined;
+    if (screenshotData?.base64) {
+      setScreenshotBase64(screenshotData.base64);
+    }
+  }, [screenshot]);
```
[the remark this answers is not in the document]
Comment (quanru) [reply packages/visualizer/src/component/blackboard/index.tsx]: Fixed in this PR. The unsafe type casting has been replaced with proper type guards that handle multiple screenshot formats:

1. **ScreenshotItem instance** (runtime): Checks for `getData` method and calls it asynchronously
2. **Serialized inline format** (`{ base64: "..." }`): Extracts the base64 string directly
3. **undefined/null**: Handles gracefully

The updated implementation at lines 107-143 now:
- Removes the `as unknown as` cast
- Uses proper type guards with `typeof` and `in` checks
- Handles async data loading with error handling
- Supports both runtime and serialized screenshot formats

This ensures the component works correctly regardless of whether screenshots are ScreenshotItem instances or serialized objects.

File: packages/core/src/screenshot-item.ts [hunk 5]
```diff
@@ -1,64 +1,92 @@
...
+      // Rollback: attempt to remove data from new provider
+      try {
+        await newProvider.delete(newId);
+      } catch {
+        // Rollback failed - data may exist in both providers
+        // Caller should handle cleanup if needed
+      }
+      throw error;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (quanru) [reply packages/core/src/screenshot-item.ts]: Fixed in this PR. The method now has comprehensive JSDoc documentation that clearly addresses this concern:

**Documentation added (lines 46-71):**
- **IMPORTANT** warning that the method returns a NEW instance
- Explicit statement: "The original instance becomes invalid after successful migration, as its data is deleted from the old provider"
- Clear instruction: "Always use the returned instance and discard the old one"
- Migration process explained step-by-step
- Error handling and rollback behavior documented
- Usage example showing correct pattern

**Why returning new instance (not mutating):**
This design is intentional for immutability and safety:
1. Makes the migration atomic - either fully succeeds or fully fails
2. Prevents accidental use of invalid state (old instance has deleted data)
3. Follows functional programming principles
4. Makes error handling clearer (old instance remains valid if migration fails)

The documentation now makes it impossible to miss that callers must use the returned instance.

File: packages/web-integration/src/static/static-page.ts [hunk 6]
```diff
@@ -165,7 +166,14 @@ export default class StaticPage implements AbstractInterface {
...
   async getContext(): Promise<UIContext> {
-    return this.uiContext;
+    // Lazy load ScreenshotItem if needed
+    if (!this.uiContext.screenshot && this.uiContext.screenshotBase64) {
+      const { ScreenshotItem } = await import('@midscene/core');
+      this.uiContext.screenshot = await ScreenshotItem.create(
+        this.uiContext.screenshotBase64,
+      );
+    }
+    return this.uiContext as UIContext;
```
[the remark this answers is not in the document]
Comment (quanru) [reply packages/web-integration/src/static/static-page.ts]: This comment is based on a misunderstanding of how `ScreenshotItem.create()` works.

**The code is actually correct:**

Looking at `ScreenshotItem.create()` implementation (screenshot-item.ts:20-26):
```typescript
static async create(
  base64: string,
  provider: StorageProvider = new MemoryStorage(),
): Promise<ScreenshotItem> {
  const id = await provider.store(base64);  // ✅ Data IS stored here
  return new ScreenshotItem(id, provider);
}
```

**What actually happens:**
1. `ScreenshotItem.create(screenshotBase64)` is called
2. A new `MemoryStorage()` is created (default parameter)
3. **`provider.store(base64)` is called**, which stores the data in MemoryStorage and returns an ID
4. A new `ScreenshotItem` is created with that ID and the storage provider
5. When `getData()` is called later, it retrieves the data from the MemoryStorage using the ID

**The data IS stored in MemoryStorage.** The concern that "data was never stored" is incorrect. The `store()` method on line 24 ensures the base64 data is persisted in the storage provider before returning.

**Review by yuyutaotao — APPROVED** [review]
